pkg/virt-launcher/virtwrap/converter/converter.go

```diff
@@ -850,6 +850,44 @@ func Convert_v1_Rng_To_api_Rng(_ *v1.Rng, rng *api.Rng, c *ConverterContext) err
 	return nil
 }
 
+func Convert_v1_Usbredir_To_api_Usbredir(vmi *v1.VirtualMachineInstance, domainDevices *api.Devices, c *ConverterContext) (bool, error) {
+	clientDevices := vmi.Spec.Domain.Devices.ClientDevices
+
+	// Default is to have USB Redirection disabled
+	if clientDevices == nil || len(clientDevices) <= 0 {
+		return false, nil
+	}
+
+	numberOfDevices := len(clientDevices)
+	redirectDevices := make([]api.Redirected, numberOfDevices)
+	if numberOfDevices > v1.UsbClientDeviceMaxNumberOf {
+		err := fmt.Errorf("Number of USB devices (%v) exceeds the maximum allowed (%v)", numberOfDevices, v1.UsbClientDeviceMaxNumberOf)
+		return false, err
+	}
+
+	// Check if name is unique
+	names := make(map[string]int)
+	for i, dev := range clientDevices {
+		if _, exists := names[dev.Usb.Name]; exists {
+			err := fmt.Errorf("Name of USB devices must be unique. The Name %v is repeated.", dev.Usb.Name)
+			return false, err
+		}
+		names[dev.Usb.Name] = i
+
```

Comment:
I see that you've already implemented this validation in `vmi-create-admitter.go` (line 2031).
I think that if this validation is being done at the manifest level there is no need to re-validate it here.

Reply:
It is true that we are double-checking (and testing) limits outside of converter.go scope. I'll rework this bit and probably remove two tests then:
- `"should reject usb redirection there is devices with the same name"`
- `"should not enable usb redirection when numberOfDevices > limit"`

And keeping only these two:
- `"should enable usb redirection when number of USB client devices > 0`
- `"should not enable usb redirection when numberOfDevices == 0"`